### PR TITLE
Odin:refactor preprocessor into flex, add macro and inline includes

### PR DESCRIPTION
**Pull request by jeanlego**
#### Description
This PR gets rid of the hand written preprocessor and use Flex states to do processing.
Also adds macro ( `define ADD(a,b) a + b ) and does recursive macro (depth limited to 256 for safety)

#### Related Issue
N/A

#### Motivation and Context
Macro was unavailable and the preprocessor was stripping away context for error reporting.

#### How Has This Been Tested?
pre_commit

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: ODIN_II/SRC/odin_error.cpp [hunk 1]
```diff
@@ -19,28 +23,38 @@ const char *odin_error_STR[] =
...
-	if (file >= 0 && file < configuration.list_of_file_names.size())
-		fprintf(stderr," (File: %s)", configuration.list_of_file_names[file].c_str());
+	if(delayed_errors)
+	{
+		error_message(PARSE_ERROR, -1, -1, "Parser found (%d) errors in your syntax, exiting", delayed_errors);
+	}
 }
 
-static void print_line_number(long line_number)
+static std::string make_marker_from_str(std::string str, int collumn)
```
Comment (ademmings) [inline ODIN_II/SRC/odin_error.cpp]: column

File: ODIN_II/SRC/odin_error.cpp [hunk 2]
```diff
@@ -19,28 +23,38 @@ const char *odin_error_STR[] =
...
+			str[i] = ' ';
+		}
+	}
+
+	str += "^~~~";
+	return str;
 }
 
-static void print_culprit_line(long line_number, long file)
+static void print_culprit_line(long collumn, long line_number, long file)
```
Comment (ademmings) [inline ODIN_II/SRC/odin_error.cpp]: column

File: ODIN_II/SRC/odin_error.cpp [hunk 3]
```diff
@@ -56,16 +70,21 @@ static void print_culprit_line(long line_number, long file)
...
+		if(collumn > 0)
+		{
+			delayed_errors++;
+			fprintf(stderr,"%s\n", make_marker_from_str(culprit_line, collumn).c_str());
+		}
 	}
 }
 
-void _log_message(odin_error error_type, long line_number, long file, bool soft_error, const char *function_file_name, long function_line, const char *function_name, const char *message, ...)
+void _log_message(odin_error error_type, long collumn, long line_number, long file, bool soft_error, const char *function_file_name, long function_line, const char *function_name, const char *message, ...)
```
Comment (ademmings) [inline ODIN_II/SRC/odin_error.cpp]: column

File: ODIN_II/SRC/odin_ii.cpp [hunk 4]
```diff
@@ -193,15 +197,22 @@ static ODIN_ERROR_CODE synthesize_verilog()
...
+		zero_string = vtr::strdup(ZERO_GND_ZERO);
+		pad_string = vtr::strdup(ZERO_PAD_ZERO);
+
+		printf("--------------------------------------------------------------------\n");
+		printf("Welcome to ODIN II version 0.1 - the better High level synthesis tools++ targetting FPGAs (mainly VPR)\n");
+		printf("Email: jamieson.peter@gmail.com and ken@unb.ca for support issues\n\n");
+	}
+	catch(vtr::VtrError& vtr_error) 
+	{
+		printf("Odin Failed Reading The command line argumenrs %s with exit code%d\n", vtr_error.what(), ERROR_INITIALIZATION);
```
Comment (ademmings) [inline ODIN_II/SRC/odin_ii.cpp]: arguments

File: ODIN_II/SRC/odin_util.cpp [hunk 5]
```diff
@@ -104,7 +104,7 @@ void assert_supported_file_extension(std::string input_file, int file_number)
 			supported_extension_list += file_extension_supported_STR[i];
 		}
 
-		possible_error_message(ARG_ERROR, -1, file_number, 
+		possible_error_message(ARG_ERROR, line_number, file_number, 
 			"File (%s) has an unsupported extension (%s), Odin only support { %s }",
```
Comment (ademmings) [inline ODIN_II/SRC/odin_util.cpp]: supports (not related but still good to fix typos while we're here)

File: ODIN_II/SRC/verilog_flex.l [hunk 6]
```diff
@@ -22,218 +22,733 @@ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
...
+#define YY_USER_ACTION {my_yycolumn = current_yycolumn; current_yycolumn = yyleng;}
+
 /* the define below helps with watching the parser go token by token */
-#define PRINT_TOKEN {printf("%d %s\n", yylineno, yytext);}
-#define UNSUPPORTED_TOKEN {printf("Unsuported token::"); PRINT_TOKEN }
-#define MP {if (to_view_parse) PRINT_TOKEN}
+#define UNSUPPORTED_TOKEN 	{ delayed_error_message(PARSE_ERROR, my_yycolumn, my_yylineno, current_parse_file, "%s", "Unsuported token"); }
+
+#define _STATE_TOP(str)		{ BEGIN(_state_top(str)); }
+#define POP_STATE()			{ _pop_state(); _STATE_TOP("Poped to "); }
```
Comment (ademmings) [inline ODIN_II/SRC/verilog_flex.l]: popped

File: ODIN_II/SRC/verilog_flex.l [hunk 7]
```diff
@@ -22,218 +22,733 @@ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
...
+#define PUSH_STATE(state)	{ _push_state(state); _STATE_TOP("Pushed to ");}
+#define CHANGE_STATE(state) { _pop_state(); _push_state(state); _STATE_TOP("Switched to"); }
```
Comment (ademmings) [inline ODIN_II/SRC/verilog_flex.l]: space after "to"? (like PUSH_STATE and POP_STATE)

**Comment by jeanlego** [comment]
@ademmings should be good to go :+1:

**Pull request merged**